src/front/wgsl/lexer.rs

```diff
@@ -13,47 +13,328 @@ fn consume_any(input: &str, what: impl Fn(char) -> bool) -> (&str, &str) {
     input.split_at(pos)
 }
 
+/// Tries to skip a given prefix in the input string.
+/// Returns whether the prefix was present and could therefore be skipped,
+/// the remaining str and the number of *bytes* skipped.
+fn try_skip_prefix<'a, 'b>(input: &'a str, prefix: &'b str) -> (bool, &'a str, usize) {
+    if let Some(input_without_prefix) = input.strip_prefix(prefix) {
+        (true, input_without_prefix, prefix.len())
+    } else {
+        (false, input, 0)
+    }
+}
+
+enum NLDigitState {
+    Nothing,
+    LeadingZero,
+    DigitBeforeDot,
+    OnlyDot,
+    DigitsThenDot,
+    DigitAfterDot,
+    Exponent,
+    SignAfterExponent,
+    DigitAfterExponent,
+}
+
+struct NumberLexerState {
+    minus: bool,
+    hex: bool,
+    leading_zeros: usize,
+    digit_state: NLDigitState,
+    uint_suffix: bool,
+}
+
+impl NumberLexerState {
+    // TODO: add proper error reporting, possibly through try_into_token function returning Result
+
+    pub fn is_valid_number(&self) -> bool {
+        match self {
+            Self {
+                minus: false, // No negative zero for integers.
+                hex,
+                leading_zeros,
+                digit_state: NLDigitState::LeadingZero,
+                ..
+            } => *hex || *leading_zeros == 1, // No leading zeros allowed in non-hex integers, "0" is always allowed.
+            Self {
+                minus,
+                hex,
+                leading_zeros,
+                digit_state: NLDigitState::DigitBeforeDot,
+                uint_suffix,
+            } => {
+                (*hex || *leading_zeros == 0) // No leading zeros allowed in non-hex integers.
+                                              // In this state the number has non-zero digits,
+                                              // i.e. it is not just "0".
+                    && (*minus ^ *uint_suffix) // Either a negative number, or and unsigned integer, not both.
+            }
+            Self {
+                digit_state:
+                    NLDigitState::DigitsThenDot
+                    | NLDigitState::DigitAfterDot
+                    | NLDigitState::DigitAfterExponent,
+                uint_suffix: false, // These states are only valid for floating point numbers.
+                ..
+            } => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_float(&self) -> bool {
+        match self {
+            Self {
+                digit_state:
+                    NLDigitState::DigitsThenDot
```

Comment:
fancy pattern matching! I wonder if it breaks our MSRV

Reply:
Ah yes, I kinda forgot about checking for MSRV compatibility. I just built the code with 1.43 and indeed [or-patterns](https://github.com/rust-lang/rust/issues/54883) and [str_strip](https://github.com/rust-lang/rust/issues/67302) are too new

Reply:
str_strip has a minimum version of 1.45 and or-patterns have 1.53 as the minimum for stable.
str_strip usage I could easily replace, but for removing the or-patterns I'd probably have to switch to if-else-if-else chains,
which may make the code harder to read still.
I assume bumping to 1.53 is not something you wanna do if avoidable, correct @kvark?

Reply:
Actually, it seems I can replace the or-patterns relatively easily after all, the only real downside is having to use match arm guards and giving up the exhaustiveness check this way (requiring the addition of a fallback arm).